test-lit/Fir/fir-ops.fir

```diff
@@ -0,0 +1,458 @@
+// CHECK-LABEL: func @dvd3() -> !fir.type<derived3{f:f32}>
+// CHECK-LABEL: func @dvd4() -> !fir.type<derived4(p:i8){f:f32}>
+func @dvd1() -> !fir.type<derived1>
+func @dvd2() -> !fir.type<derived2(p:i32)>
+func @dvd3() -> !fir.type<derived3{f:f32}>
+func @dvd4() -> !fir.type<derived4(p:i8){f:f32}>
+
+// FIR array types
+// CHECK-LABEL: func @arr1() -> !fir.array<10xf32>
+// CHECK-LABEL: func @arr2() -> !fir.array<10x10xf32>
+// CHECK-LABEL: func @arr3() -> !fir.array<?xf32>
+// CHECK-LABEL: func @arr4() -> !fir.array<10x?xf32>
+// CHECK-LABEL: func @arr5() -> !fir.array<?x?xf32>
+// CHECK-LABEL: func @arr6() -> !fir.array<*:f32>
+func @arr1() -> !fir.array<10xf32>
+func @arr2() -> !fir.array<10x10xf32>
+func @arr3() -> !fir.array<?xf32>
+func @arr4() -> !fir.array<10x?xf32>
+func @arr5() -> !fir.array<?x?xf32>
+func @arr6() -> !fir.array<*:f32>
+
+// FIR pointer-like types
+// CHECK-LABEL: func @mem1() -> !fir.ref<i32>
+// CHECK-LABEL: func @mem2() -> !fir.ptr<i32>
+// CHECK-LABEL: func @mem3() -> !fir.heap<i32>
+// CHECK-LABEL: func @mem4() -> !fir.ref<() -> ()>
+func @mem1() -> !fir.ref<i32>
+func @mem2() -> !fir.ptr<i32>
+func @mem3() -> !fir.heap<i32>
+func @mem4() -> !fir.ref<() -> ()>
+
+// FIR box types (descriptors)
+// CHECK-LABEL: func @box1() -> !fir.box<!fir.array<?xf32>>
+// CHECK-LABEL: func @box2() -> !fir.boxchar<2>
+// CHECK-LABEL: func @box3() -> !fir.boxproc<(i32, i32) -> i64>
+// CHECK-LABEL: func @box4() -> !fir.box<!fir.type<derived3{f:f32}>>
+func @box1() -> !fir.box<!fir.array<?xf32>>
+func @box2() -> !fir.boxchar<2>
+func @box3() -> !fir.boxproc<(i32, i32) -> i64>
+func @box4() -> !fir.box<!fir.type<derived3{f:f32}>>
+
+// FIR misc. types
+// CHECK-LABEL: func @oth1() -> !fir.dims<1>
+// CHECK-LABEL: func @oth2() -> !fir.field
+// CHECK-LABEL: func @oth3() -> !fir.tdesc<!fir.type<derived5{f1:f32,f2:f32}>>
+func @oth1() -> !fir.dims<1>
+func @oth2() -> !fir.field
+func @oth3() -> !fir.tdesc<!fir.type<derived5{f1:f32,f2:f32}>>
+
+// Fortran SUBROUTINE and FUNCTION
+// CHECK-LABEL: func @print_index3(index, index, index)
+// CHECK-LABEL: func @user_i64(i64)
+// CHECK-LABEL: func @user_tdesc(!fir.tdesc<!fir.type<x>>)
+func @print_index3(index, index, index)
+func @user_i64(i64)
+func @user_tdesc(!fir.tdesc<!fir.type<x>>)
+
+// expect the void return to be omitted
+// CHECK-LABEL: func @store_tuple(tuple<!fir.type<qq1{f1:i32}>>)
+func @store_tuple(tuple<!fir.type<qq1{f1:i32}>>) -> ()
+
+// CHECK-LABEL: func @get_method_box() -> !fir.box<!fir.type<derived3{f:f32}>>
+// CHECK-LABEL: func @method_impl(!fir.box<!fir.type<derived3{f:f32}>>)
+func @get_method_box() -> !fir.box<!fir.type<derived3{f:f32}>>
+func @method_impl(!fir.box<!fir.type<derived3{f:f32}>>)
+
+// CHECK-LABEL: func @nop()
+func @nop()
+
+// CHECK-LABEL: func @get_func() -> (() -> ())
+func @get_func() -> (() -> ())
+
+// CHECK-LABEL: @instructions
+func @instructions() {
+  // CHECK: %[[A0:.*]] = fir.alloca !fir.array<10xi32>
```

Comment:
Not sure if you all are aware, but there is a script to help generate CHECK lines:
https://github.com/llvm/llvm-project/blob/master/mlir/utils/generate-test-checks.py

Reply:
Thanks. I'm making a note of this.